python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/updated_scenarios/basic_scenarios.py

```diff
@@ -0,0 +1,290 @@
+from dagster import AssetSpec, AutoMaterializeRule
+from dagster._core.definitions.auto_materialize_rule import ParentUpdatedRuleEvaluationData
+
+from ..asset_daemon_scenario import (
+    AssetDaemonScenario,
+    AssetDaemonScenarioState,
+    AssetRuleEvaluationSpec,
+)
+from ..asset_daemon_scenario_states import (
+    diamond,
+    one_asset,
+    one_asset_depends_on_two,
+    three_assets_in_sequence,
+    two_assets_depend_on_one,
+    two_assets_in_sequence,
+)
+from ..base_scenario import run_request
+
+basic_scenarios = [
+    AssetDaemonScenario(
+        id="one_asset_never_materialized",
+        initial_state=one_asset.with_all_eager(),
+        scenario=lambda state: state.evaluate_tick()
```

Comment:
Putting lambdas everywhere feels a little awkward. Did you consider an approach where we do `AssetDaemonScenario(id=..., initial_state=...).evaluate_tick().assert_requested_runs(...)`?

Reply:
The lambda is there to avoid the scenario from executing anything at import time. I briefly considered some scheme where all the methods on AssetDaemonScenario resulted in appending to a list of "pending actions" or something on the scenario, then having a final "evaluate_all()" function at the end which would be called within the test, but that seemed like more trouble than it was worth

Reply:
> I briefly considered some scheme where all the methods on AssetDaemonScenario resulted in appending to a list of "pending actions" or something on the scenario, then having a final "evaluate_all()" function at the end which would be called within the test, but that seemed like more trouble than it was worth

Yeah that's what I was imagining. If a list of pending actions is too annoying, then implementing it with closures could be easier?

```python
class AssetDaemonScenario:
    state: AssetDaemonScenarioState
    execute_fn: Callable[[AssetDaemonScenarioState, DagsterInstance], None]

    def execute(self, instance: DagsterInstance) -> AssetDaemonScenarioState:
        self.execute_fn(self.state, instance)

    def evaluate_tick(self) -> AssetDaemonScenario:
        def _execute_and_evaluate_tick(state, instance) -> AssetDaemonScenarioState:
            state = self.execute_fn(state, instance)

            AssetDaemonContext(instance=instance, asset_graph=state.asset_graph, ...).evaluate()
            return state._replace(...)

        return AssetDaemonScenario(self.state, _execute_and_evaluate_tick)
```

If that's too annoying I think we should at least rename `scenario` to `execute_fn` or something.

Reply:
mm yeah that seems good -- although just for my own sanity I'll implement that in a separate PR upstack (and leave the "scenario" name alone for now)

Reply:
I thought about this a little more while working on auto-materialize perf stuff, and one thing to think about how useful stack traces are. E.g. with the lambda approach, if code hits an error, are we more likely to be able to see that it occurred during `state.evaluate_tick` on X line? If so, maybe it's worth keeping the lambdas around.

Reply:
ah yeah I did throw a bit of time into this on friday and the stack traces were quite rough in my attempts. The other downside is that it's much easier to mess up writing a function in this style (i.e. accidentally type "self" instead of "state" or something of the sort, and debugging an issue of that sort isn't the easiest.

Considering the benefit here is fairly small (just a somewhat sleeker interface), I'm leaning towards keeping the lambdas (and I'll update the `scenario` name) if that sounds good to you

Reply:
Cool, sounds good